src/core/trulens/experimental/otel_tracing/core/instrument.py

```diff
@@ -137,3 +157,69 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return inner_decorator
+
+
+class App(core_app.App):
+    # For use as a context manager.
+    def __enter__(self):
+        logger.debug("Entering the OTEL app context.")
+
+        # Note: This is not the same as the record_id in the core app since the OTEL
+        # tracing is currently separate from the old records behavior
+        otel_record_id = str(uuid.uuid4())
+
+        tracer = trace.get_tracer_provider().get_tracer(TRULENS_SERVICE_NAME)
+
+        # Calling set_baggage does not actually add the baggage to the current context, but returns a new one
+        # To avoid issues with remembering to add/remove the baggage, we attach it to the runtime context.
+        self.tokens.append(
+            context_api.attach(
+                set_baggage(SpanAttributes.RECORD_ID, otel_record_id)
+            )
+        )
+        self.tokens.append(
+            context_api.attach(set_baggage(SpanAttributes.APP_ID, self.app_id))
+        )
+
+        # Use start_as_current_span as a context manager
+        self.span_context = tracer.start_as_current_span("root")
+        root_span = self.span_context.__enter__()
+
+        logger.debug(str(get_baggage(SpanAttributes.RECORD_ID)))
+
+        # Set general span attributes
+        root_span.set_attribute("kind", "SPAN_KIND_TRULENS")
+        root_span.set_attribute("name", "root")
+        root_span.set_attribute(
+            SpanAttributes.SPAN_TYPE, SpanAttributes.SpanType.RECORD_ROOT
+        )
+        root_span.set_attribute(SpanAttributes.APP_ID, self.app_id)
+        root_span.set_attribute(SpanAttributes.RECORD_ID, otel_record_id)
+
+        # Set record root specific attributes
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.APP_NAME, self.app_name
+        )
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.APP_VERSION, self.app_version
+        )
+        root_span.set_attribute(SpanAttributes.RECORD_ROOT.APP_ID, self.app_id)
+        root_span.set_attribute(
+            SpanAttributes.RECORD_ROOT.RECORD_ID, otel_record_id
+        )
+
+        return root_span
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        remove_baggage(SpanAttributes.RECORD_ID)
+        remove_baggage(SpanAttributes.APP_ID)
+
+        logger.debug("Exiting the OTEL app context.")
+
+        while len(self.tokens) > 0:
+            # Clearing the context once we're done with this root span.
+            # See https://github.com/open-telemetry/opentelemetry-python/issues/2432#issuecomment-1593458684
+            context_api.detach(self.tokens.pop())
+
+        if self.span_context:
```

Comment:
would this ever not be true?

Reply:
I doubt it, but didn't think it would hurt to include it, if nothing, at least for the type checking haha